controlplane/kubeadm/internal/filters.go

```diff
@@ -41,65 +40,73 @@ import (
+		logReasons = append(logReasons, fmt.Sprintf("Machine version %q is not equal to KCP version %q", machineVersion, kcp.Spec.Version))
+		conditionReasons = append(conditionReasons, "spec.version is not up to date")
 	}
 
 	reason, matches, err := matchesKubeadmBootstrapConfig(machineConfigs, kcp, machine)
 	if err != nil {
-		return "", false, errors.Wrapf(err, "failed to match Machine spec")
+		return false, nil, nil, errors.Wrapf(err, "failed to match Machine spec")
 	}
 	if !matches {
-		mismatchReasons = append(mismatchReasons, reason)
+		logReasons = append(logReasons, reason)
+		conditionReasons = append(conditionReasons, "spec.bootstrap is not up to date")
 	}
 
 	if reason, matches := matchesTemplateClonedFrom(infraConfigs, kcp, machine); !matches {
-		mismatchReasons = append(mismatchReasons, reason)
+		logReasons = append(logReasons, reason)
+		conditionReasons = append(conditionReasons, "spec.infrastructureRef is not up to date")
 	}
 
-	if len(mismatchReasons) > 0 {
-		return strings.Join(mismatchReasons, ","), false, nil
+	if len(logReasons) > 0 || len(conditionReasons) > 0 {
+		return false, logReasons, conditionReasons, nil
 	}
 
-	return "", true, nil
+	return true, nil, nil, nil
 }
 
 // NeedsRollout checks if a Machine needs to be rolled out and returns the reason why.
-func NeedsRollout(reconciliationTime, rolloutAfter *metav1.Time, rolloutBefore *controlplanev1.RolloutBefore, infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane, machine *clusterv1.Machine) (string, bool, error) {
-	rolloutReasons := []string{}
+func NeedsRollout(reconciliationTime, rolloutAfter *metav1.Time, rolloutBefore *controlplanev1.RolloutBefore, infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane, machine *clusterv1.Machine) (bool, []string, []string, error) {
+	logReasons := []string{}
+	conditionReasons := []string{}
 
 	// Machines whose certificates are about to expire.
 	if collections.ShouldRolloutBefore(reconciliationTime, rolloutBefore)(machine) {
-		rolloutReasons = append(rolloutReasons, "certificates will expire soon, rolloutBefore expired")
+		logReasons = append(logReasons, "certificates will expire soon, rolloutBefore expired")
+		conditionReasons = append(conditionReasons, "certificates will expire soon")
 	}
 
 	// Machines that are scheduled for rollout (KCP.Spec.RolloutAfter set,
 	// the RolloutAfter deadline is expired, and the machine was created before the deadline).
 	if collections.ShouldRolloutAfter(reconciliationTime, rolloutAfter)(machine) {
-		rolloutReasons = append(rolloutReasons, "rolloutAfter expired")
+		logReasons = append(logReasons, "rolloutAfter expired")
+		conditionReasons = append(conditionReasons, "kcp.spec.rolloutAfter expired")
```

Comment:
```suggestion
		conditionReasons = append(conditionReasons, "KCP spec.rolloutAfter expired")
```
(probably the way to write this that is consistent with other places)

Reply:
Aligned to KubeadmControlPlane (we should not assume users knows the KCP abbreviation)